File: 01-05-2024/Arrange Consonants and Vowels.py
#User function Template for python3

"""
# Node Class

class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

"""


#Your Code Start
class Solution:
    #Function to reverse a linked list.
    def arrangeCV(self, head):
        # If the head is None or there's only one node, return the head as it is
        if head is None or head.next is None:
            return head
        
        # Initialize pointers for vowels and consonants
        vowel_head = vowel_tail = None
        consonant_head = consonant_tail = None

        current = head
        while current:
            if current.data.lower() in ['a', 'e', 'i', 'o', 'u']:
                # If the current node is a vowel
                if vowel_head is None:
                    vowel_head = vowel_tail = current
                else:
                    vowel_tail.next = current
                    vowel_tail = current
            else:
                # If the current node is a consonant
                if consonant_head is None:
                    consonant_head = consonant_tail = current
                else:
                    consonant_tail.next = current
                    consonant_tail = current
            # Move to the next node
            current = current.next

        # If there are no vowels, return the original list
        if vowel_head is None:
            return consonant_head

        # Concatenate the two lists
        vowel_tail.next = consonant_head
        # Set the next of the last node in consonant list to None
        consonant_tail.next = None

        # Return the head of the updated linked list
        return vowel_head
#Your Code End

#{ 
 # Driver Code Starts
# Node Class
class Node:

    def __init__(self, val):
        self.data = val
        self.next = None


# Linked List Class
class Linked_List:

    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, val):
        if self.head is None:
            self.head = Node(val)
            self.tail = self.head
        else:
            self.tail.next = Node(val)
            self.tail = self.tail.next


def printList(head):
    tmp = head
    while tmp:
        print(tmp.data, end=' ')
        tmp = tmp.next
    print()


if __name__ == '__main__':
    for i in range(int(input())):
        n = int(input())
        arr = [str(x) for x in input().split()]

        lis = Linked_List()
        for i in arr:
            lis.insert(i)

        newHead = Solution().arrangeCV(lis.head)
        printList(newHead)

# } Driver Code Ends

File: 01-05-2024/Arrange Consonants and Vowels.java
//{ Driver Code Starts
/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class Node {
    char data;
    Node next;
    
    public Node(char data){
        this.data = data;
        next = null;
    }
}

class GFG {
	public static void main (String[] args) {
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		while(t-- > 0){
		    int n = sc.nextInt();
		    Node head = null, tail = null;
		    
		    char head_c = sc.next().charAt(0);
		    head = new Node(head_c);
		    tail = head;
		    
		    while(n-- > 1){
		        tail.next = new Node(sc.next().charAt(0));
		        tail = tail.next;
		    }
		    
		    Solution obj = new Solution();
		    //show(head);
		    show(obj.arrangeCV(head));
		    
		}
	}
	
	public static void po(Object o){
	    System.out.println(o);
	}
	
	public static void show(Node head){
        while(head != null){
            System.out.print(head.data+" ");
            head = head.next;
        }
        System.out.println();
    }
}

// } Driver Code Ends


/*
Structure of node class is:
class Node {
    char data;
    Node next;
    
    public Node(char data){
        this.data = data;
        next = null;
    }
}
*/
//Your Code Start
class Solution {
    
    public Node arrangeCV(Node head) {
        // If the head is null or there's only one node, return the head as it is
        if (head == null || head.next == null) {
            return head;
        }
        
        // Initialize pointers for vowels and consonants
        Node vowelHead = null, vowelTail = null;
        Node consonantHead = null, consonantTail = null;
        
        // Traverse the linked list
        while (head != null) {
            // If the current character is a vowel
            if (isVowel(head.data)) {
                // If the vowel list is empty, initialize it
                if (vowelHead == null) {
                    vowelHead = vowelTail = head;
                } else {
                    // Append the current node to the vowel list
                    vowelTail.next = head;
                    vowelTail = head;
                }
            } else {
                // If the consonant list is empty, initialize it
                if (consonantHead == null) {
                    consonantHead = consonantTail = head;
                } else {
                    // Append the current node to the consonant list
                    consonantTail.next = head;
                    consonantTail = head;
                }
            }
            // Move to the next node
            head = head.next;
        }
        
        // Connect the vowel list with the consonant list
        if (vowelHead != null) {
            vowelTail.next = consonantHead;
        }
        
        // If the consonant list is not empty, set its tail's next to null
        if (consonantTail != null) {
            consonantTail.next = null;
        }
        
        // Return the head of the rearranged linked list
        return vowelHead != null ? vowelHead : consonantHead;
    }
    
    // Function to check if a character is a vowel
    private boolean isVowel(char ch) {
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }
}
//Your Code End
